commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+import org.infinispan.commons.api.functional.EntryView.ReadWriteEntryView;
+import org.infinispan.commons.api.functional.EntryView.WriteEntryView;
+import org.infinispan.commons.api.functional.Listeners.ReadWriteListeners;
+import org.infinispan.commons.api.functional.Listeners.WriteListeners;
+import org.infinispan.commons.util.CloseableIterator;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Top level functional map interface offering common functionality for the
+ * read-only, read-write, and write-only operations that can be run against a
+ * functional map.
+ *
+ * DESIGN RATIONALES:
+ * <ul>
+ *    <li>Originally, I tried to come up with a single functional map interface
+ *    that would encompass read-only, read-write and write-only operations, but
+ *    it felt quite bloated. By separating each major type of operations to its
+ *    own interface, it becomes easier to figure out which are all the read-only
+ *    operations, which are write-only...etc, which also helps the user quickly
+ *    see what they can do without being obstructed by other operation types.
+ *    </li>
+ *    <li>The conscious decision to separate read-only, write-only and
+ *    read-write interfaces helps massively with type safety. So, if a user
+ *    gets a read-only map, it can't write to it by mistake since no such
+ *    APIs are exposed. The same happens with write-only maps, the user can
+ *    only write and cannot make the mistake of reading from the entry view
+ *    because read operations are not exposed. The previous design did not
+ *    have this level of type safety and hence it relied on runtime exceptions
+ *    to catch these wrong combinations.</li>
```

Comment:
Not sure you fully understood what I'm trying to do here. Instead of having a Cache or similar interface with all the operations in in, including read-only operations such as get(), I've separated here all the operations that are read-only, vs those that are only write-only, and read-write. This makes it easier for the reader to know the implications of each group of operations, e.g. read-only operations never acquire locks, read-write operations are used for conditional operations or when trying to return previous values and hence require previous values to be brought back from persistence or another node, and write-only operations never require previous values....etc. 

So, ReadOnlyMap is just a read-only view of the FunctionalMap, for when you want to only call read-only operations. Users call read-only operations, such as get(), the only difference is that these are now more clearly separated.

I think the separation makes a lot of sense and it's easier to understand the implications of each group set of operations, as opposed to having a kitchen sink interface (our Cache or JSR-107 cache) with all the operations you can think of thrown in.